tests/integration/records/table_validator.py

```diff
@@ -62,9 +62,15 @@ def validate(self,
         self.validate_data_values(schema_name, table_name)
 
     def validate_data_types(self, schema_name: str, table_name: str) -> None:
-        columns = self.target_db_engine.dialect.get_columns(self.target_db_engine,
-                                                            table_name,
-                                                            schema=schema_name)
+        if self.target_db_engine.driver == 'pymysql':
+            conn = self.target_db_engine.connect()
+            columns = self.target_db_engine.dialect.get_columns(conn,
+                                                                table_name,
+                                                                schema=schema_name)
+        else:
+            columns = self.target_db_engine.dialect.get_columns(self.target_db_engine,
+                                                                table_name,
+                                                                schema=schema_name)
```

Comment:
This is so weird to me, does the get_columns method take different types of objects depending on what the driver is????

Reply:
From the sqlalchemy documentation, get_columns says it always takes a connection as the first arg.
https://docs.sqlalchemy.org/en/14/core/internals.html#sqlalchemy.engine.Dialect.get_columns.

I'm sure you've already tried this, but we can't just pass the `conn` object to both?

Is this records-mover's fault for storing inconsistent types in its wrappers around sqlalchemy?

Reply:
```
class DBDriver(metaclass=ABCMeta):
    def __init__(self,
                 db: Union[sqlalchemy.engine.Engine,
                           sqlalchemy.engine.Connection], **kwargs) -> None:
```
well this is weird

Reply:
Alright so I think the reason for this is that in some cases we want to instantiate a db_driver in the context of a transaction, as in this example from records/mover/records/prep_and_load.py
```python
    with tbl.db_engine.begin() as db:
        # This second transaction ensures the table has been created
        # before non-transactional statements like Redshift's COPY
        # take place.  Otherwise you'll get an error like:
        #
        #  Cannot COPY into nonexistent table
        driver = tbl.db_driver(db)
        try:
            import_count = load(driver)
        except load_exception_type:
            if not tbl.drop_and_recreate_on_load_error:
                raise
            reset_before_reload()
            with tbl.db_engine.begin() as db:
                driver = tbl.db_driver(db)
                prep.prep(schema_sql=schema_sql,
                          driver=driver,
                          existing_table_handling=ExistingTableHandling.DROP_AND_RECREATE)
                import_count = load(driver)
        return MoveResult(move_count=import_count, output_urls=None)
```
`begin()` starts a transaction and returns a connection object - naming it a "db" is pretty confusing.

Anyway, other times we don't care about transactions and don't create the connection object beforehand. I'm kind of confused about this, but sqlalchemy had this feature "connection-less execution" which basically let you execute stuff on an `engine` directly without needing a `connection` first, but they've been phasing it out: https://docs.sqlalchemy.org/en/14/core/connections.html#connectionless-execution-implicit-execution. Maybe this is why the code used to work.

I think the proper solution is to split the union from `DBDriver` into an optional `connection` argument and an `Engine`. If no `connection` is given it can create one from the `Engine`. Then, when a caller wants to do something like `get_columns`, they always use the `connection` attribute of `DBDriver`, not the `Engine`. This solves a bunch of confusion caused by examples such as above, where `target_db_engine` isn't actually an engine, it's a connection.

Should we track this work separately or do it as part of this ticket?

If we choose to put this work off until later, in the short term I think we should change the above code that I commented on (as well as the other similar occurrences) to read something like
```python
if isinstance(self.target.db_engine, sqlalchemy.engine.Connection):
    connection = self.target_db_engine
else:
    connection = self.target._db_engine.connect()
get_columns(connection)
```

Reply:
Implemented this approach.